### PR TITLE
[WIP] threads review/adjustments

### PR DESCRIPTION
**Pull request by ggazzo**
I have opened that PR to show/share the modifications I made to be able to merge on rocketchat core (:

### CODE REVIEW

File: .meteor/versions [hunk 1]
```diff
@@ -226,7 +226,6 @@ rocketchat:slackbridge@0.0.1
 rocketchat:slashcommands-archive@0.0.1
 rocketchat:slashcommands-asciiarts@0.0.1
 rocketchat:slashcommands-create@0.0.1
-rocketchat:slashcommands-create-thread@0.0.1
```
Comment (mrsimpson) [inline .meteor/versions]: Did you not like the slash command? I use it frequently myself

File: packages/assistify-threading/client/createThreadMessageAction.js [hunk 2]
```diff
@@ -1,48 +1,37 @@
...
-			Meteor.call('createThreadFromMessage', message, function(error, result) {
-				if (result) {
-					// remove the hidden message from the UI - the message list is not reactive
-					Tracker.nonreactive(function() {
-						ChatMessage.remove({
-							_id: message._id,
-						});
-					});
+	Tracker.autorun(() => {
+		if (RocketChat.settings.get('Thread_from_context_menu') !== 'button') {
```
Comment (mrsimpson) [inline packages/assistify-threading/client/createThreadMessageAction.js]: much better... Kind-of-ashamed...

File: packages/rocketchat-ui-sidenav/client/sideNav.html [hunk 3]
```diff
@@ -2,9 +2,6 @@
 		<aside class="sidebar sidebar--{{sidebarViewMode}} {{#if sidebarHideAvatar}}sidebar--hide-avatar{{/if}}" role="navigation">
 			{{> sidebarHeader }}
 			{{#if loggedInUser}}
-				{{#if threadingFromSidebar}}
```
Comment (mrsimpson) [inline packages/rocketchat-ui-sidenav/client/sideNav.html]: I absolutely understand you didn't like that button - I don't like it either. But: Is there a replacement for that? I believe there should be an option to start a thread without opening the room first.

File: packages/rocketchat-ui-sidenav/client/sideNav.js [hunk 4]
```diff
@@ -68,10 +64,6 @@ Template.sideNav.events({
 	'dropped .sidebar'(e) {
 		return e.preventDefault();
 	},
-
-	'click .js-create-thread'() {
-		return FlowRouter.go('create-thread');
```
Comment (mrsimpson) [inline packages/rocketchat-ui-sidenav/client/sideNav.js]: So you left the route, as far as I can see. 👍

**Comment by ggazzo** [comment]
@mrsimpson I'm afraid I didnt finish yet, I realize I removed your "cloud tags", actually I couldnt open/find on `https://threads-qa.rocket.chat`...

I had to change some "internal things(popover/modal)" to fit the forms inside popups... and some "props names" we are trying to "normalize"... there are some bugs I didnt fix yet(remove parent thread), but I will :)...

File: .meteor/versions (hunk 1 above)
Comment (ggazzo) [reply .meteor/versions]: I will reinsert :)

**Comment by mrsimpson** [comment]
> @mrsimpson I'm afraid I didnt finish yet, 

Of course - there's a lot to do if you want to catch all the details. I "approved" it just to signal "if you want to, just merge it any time. Since you are a maintainer, I'll (have to) accept it anyway ;) "

> I realize I removed your "cloud tags", actually I couldnt open/find on `https://threads-qa.rocket.chat`...

Ah, yes. We tried a word cloud for selecting the parent channel. However, this didn't work out as we wanted to, so we thought a normal autocomplete should do as well. I have hidden it from the UI (might be only a `display: none`), but it's not a big loss if you deleted the wordcloud.

> I had to change some "internal things(popover/modal)" to fit the forms inside popups... and some "props names" we are trying to "normalize"

I still think that the DB should be responsible for compression and thus preferred `parentRoomId` instead of `prid` (see [here](https://stackoverflow.com/a/12790994) and particularly [here](https://jira.mongodb.org/browse/SERVER-863?focusedCommentId=812326&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-812326) ), but well - most of the DB prop names are cryptical as well...

> there are some bugs I didnt fix yet(remove parent thread), but I will :)...

Cascading delete *might* be hard to understand in the context of heavy threading, since the rooms are only very loosely coupled. What about an option to keep the threads or at least a warning when deleting a room with threads?

In general, it sounds as if you do have a plan - which is good ;)
Have you seen my open questions in https://github.com/RocketChat/Rocket.Chat/pull/11803#issuecomment-452229168 ?

File: packages/assistify-threading/authorization.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	];
+
+	for (const permission of permissions) {
+		if (!RocketChat.models.Permissions.findOneById(permission._id)) {
+			RocketChat.models.Permissions.upsert(permission._id, { $set: permission });
+		}
+	}
+
+	addRoomAccessValidator(function(room, user) {
+		return room.prid && canAccessRoom(Rooms.findOne(room.prid), user);
```
Comment (mrsimpson) [inline packages/assistify-threading/authorization.js]: Ah, nice. I thought about this too. This makes it optional to invite users to the thread, which was a major flaw in the current implementation.

However, we got feedback that some users shall be able to join a thread while not being able to see the parent. Is this possible?

File: packages/assistify-threading/client/views/ThreadList.js [hunk 6]
```diff
@@ -37,7 +18,7 @@ Template.ThreadList.onRendered(function() {
 			sort[this.identifier === 'd' && RocketChat.settings.get('UI_Use_Real_Name') ? 'lowerCaseFName' : 'lowerCaseName'] = /descending/.test(sortBy) ? -1 : 1;
 		}
 
-		query.parentRoomId = { $exists: true };
-		this.threadSubscriptions.set(ChatSubscription.find(query, { sort }).fetch());
-	});
+		query.prid = { $exists: true };
+		return ChatSubscription.find(query, { sort });
```
[the remark this answers is not in the document]
Comment (ggazzo) [reply packages/assistify-threading/client/views/ThreadList.js]: let me know if you remember why you choose that way :)

File: packages/assistify-threading/client/views/fieldTypeThreadLastMessageAge.html [hunk 7]
```diff
@@ -1,5 +1,7 @@
 <template name="LastMessageAge">
 	{{#with field}}
-		<button class="rc-tags__tag rc-tags--no-icon">{{lastMessageAge}}</button>
+		{{#if lastMessageAge}}
```
[the remark this answers is not in the document]
Comment (ggazzo) [reply packages/assistify-threading/client/views/fieldTypeThreadLastMessageAge.html]: I was getting `invalid date` if I dont have any reply, with this code the problem disapears 
```javascript
lastMessageTimestamp && moment(lastMessageTimestamp).format('LLL');
```
but the empty element is there.

File: packages/assistify-threading/client/views/creationDialog/CreateThread.html [hunk 8]
```diff
@@ -1,68 +1,88 @@
...
-								</div>
+	<section class="create-channel">
+		<div class="create-channel__wrapper">
+			<header class="create-channel__header">
+				<h1 class="create-channel__title">{{_ "Threading_title"}}</h1>
+				<p class="create-channel__description">{{_ "Threading_description"}}</p>
+			</header>
+			<form id="create-thread" class="create-channel__content">
+				{{#unless disabled}}
+				{{> search
```
[the remark this answers is not in the document]
Comment (ggazzo) [reply packages/assistify-threading/client/views/creationDialog/CreateThread.html]: actually I made because its very complicate to set up from zero

**Comment by mrsimpson** [comment]
@ggazzo I have started to test a bit of the current implementation.
I assume you know most of it, still I wanted to note them:

## Creation-dialogue

- [ ] the parent channel selection shows an `@` instead of a `#`
- [ ] The selection of the parent looks like a multi-select (like user selection)
- [ ] You cannot select a private group
- [ ] Once you start a thread from a message, you can still enter a new message. I don't understand this option. 
- [ ] The option to invite users to the thread is optional: All parent channel users will be allowed. This behavior should be understandable from the screen, e. g. by having a good placeholder or a help text.

## After thread creation

- [ ] Now the origin message is still there and not a system message, the it is hardly possible to determine it as a thread-link.
- [ ] Now it's a normal message, you can reply to it. This should create a reply in the thread, else it jeopardizes the structure
- [ ] When navigating via the quoted message and forth again, messages will be lost from the parent channel display (it appears empty). 
image
image
- [ ] The message from which the thread started is elementary. It should be visualized in the header of the thread so that it does not get lost when scrolling

**Comment by ThomasRoehl** [comment]
comes with RC 1.0

**Pull request closed without merging**